New JMESPATH extractor

ubikloadpack opened the pull request:
## Description

UbikLoadPack team is happy to contribute this new component that allows using JMESPATH (http://jmespath.org/) as a new technology to extract data from JSON.

Regards
Qian Qian
Philippe 


## Motivation and Context

JSON-PATH is available today but has many limits:
- As seen here https://bz.apache.org/bugzilla/show_bug.cgi?id=63623, the Java library does not offer what JSON-PATH in other implementation provides which confuses users
- JMESPATH has a specification, implementations in different languages and a test suite
- It is more powerful

## How Has This Been Tested?

JUnit Test case
Manual Testing

## Screenshots (if appropriate):

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [X] My code follows the [code style][style-guide] of this project.
- [X] I have updated the documentation accordingly.
- [X] Add elements in saveservice.properties
- [X] Remove double quotes around string values

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

pmouawad committed:
New JMES Path Extractor

pmouawad committed:
Better classes naming, drop use of Mongo for JSON parsing, better variables/constants naming, fix tests

pmouawad committed:
Remove static block
Prepare for ability to register Function

pmouawad committed:
Add checksums

vlsi commented on src/bom/build.gradle.kts:
```diff
@@ -77,6 +77,8 @@ dependencies {
         apiv("commons-lang:commons-lang")
         apiv("commons-net:commons-net")
         apiv("dnsjava:dnsjava")
+	apiv("io.burt:jmespath-core")
```
Please refrain from using tabs.
We should probably setup a checkstyle verification

pmouawad committed:
Add missing properties in saveservice.properties

pmouawad committed:
Change component label to explicitely mention JMESPath

pmouawad committed:
Fix component name

pmouawad committed:
Remove tabs

pmouawad committed:
Update FILEVERSION after modification of saveservice.properties

pmouawad committed:
Mention in changes to trigger new build

vlsi commented:
Please fix Travis failures as well

ubikloadpack commented:
> Please fix Travis failures as well

Last failure has no relation with PR, fixed by pmouawad commit.

Other ones fixed this morning

vlsi commented:
Let me clarify.
Please fix violations in JMESPathExtractor  and JMESPathExtractorGui: https://travis-ci.org/apache/jmeter/jobs/581079105#L390-L400

pmouawad committed:
Fix checkstyle

pmouawad committed:
Fix checkstyle

FSchumacher commented on src/components/src/main/java/org/apache/jmeter/extractor/json/jmespath/JMESPathExtractor.java:
```diff
@@ -0,0 +1,242 @@
+        }
+    }
+
+    public List<String> splitJson(JsonNode jsonNode) throws JsonParseException, JsonMappingException, IOException {
```
The exceptions are subclasses of `IOException`. Do we  really need to specify them here?

FSchumacher commented on src/components/src/main/java/org/apache/jmeter/extractor/json/jmespath/JMESPathExtractor.java:
```diff
@@ -0,0 +1,242 @@
+        ObjectMapper mapper = new ObjectMapper();
+        if (jsonNode.isArray()) {
+            ArrayNode arrayNode = (ArrayNode) jsonNode;
+            for (int i = 0; i < arrayNode.size(); i++) {
```
Why not use a for loop like
```
for (JsonNode element : ((ArrayNode) jsonNode)) {
    splittedJsonElements.add(mapper.writeValueAsString(element));
}
```

FSchumacher commented on src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java:
```diff
@@ -0,0 +1,237 @@
+        context.setVariables(vars);
+        vars.put("contentvar", "[\"one\"]");
+        processor.process();
+        assertThat(vars.get(REFERENCE_NAME), CoreMatchers.is(CoreMatchers.nullValue()));
```
Maybe this assert sequence could be extracted into a helper assert method like `assertValues(vars, values...)`?

FSchumacher commented on src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    @Test
+    public void testJMESPathExtractorAllElementsOneMatch() {
```
This seems to be a quite complex test, would it be better to split the tested components into simpler tests?

FSchumacher commented on src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    @Test
+    public void testJMESPathExtractorAllElementsMultipleMatches() {
```
Like the test before, would it not be better to split it into simpler tests?

pmouawad committed:
Take into account Felix Schumacher's review
Rename setMatchNumbers to setMatchNumber

pmouawad committed:
Fix checkstyle issues:
- Uncessary parenthesis
- Trailing whitespace ARGHHHH, I hate this one
- long line

pmouawad committed:
Increase coverage

pmouawad committed:
Ensure clearOldRefVars is covered

pmouawad committed:
When json node is string, strip double quote, otherwise feature is hard
to use

vlsi commented:
I see @FSchumacher has added relevant license files to the JMeter repository, however it would be nice if we could ask upstream project to include LICENSE file as `META-INF/LICENSE`.

See samples here: https://github.com/apache/jmeter/pull/469#issuecomment-502834212

@pmouawad would you please ask jmespath for shipping relevant license inside jmespath releases so we could get always up to date version of it?

PS. I've added dependency license verification to Travis as well.
